src/couch/src/couch_bt_engine_cache.erl

```diff
@@ -0,0 +1,254 @@
+        true ->
+            case should_clean(Cache) of
+                true ->
+                    ets:match_delete(Tid, {'_', 0, '_'});
+                false ->
+                    timer:sleep(wait_interval(?CLEANUP_INTERVAL_MSEC)),
+                    clean(Cache, Until)
+            end;
+        false ->
+            false
+    end.
+
+should_clean(#cache{tid = Tid, max_size = Max}) ->
+    ets:info(Tid, memory) >= Max * leave_percent() / 100.
+
+remove_dead(#cache{tid = Tid}) ->
+    All = pids(Tid),
+    Alive = sets:filter(fun is_process_alive/1, All),
+    Dead = sets:subtract(All, Alive),
+    % In OTP 27+ use sets:map/2
+    Fun = fun(Pid, _) -> ets:match_delete(Tid, {{Pid, '_'}, '_', '_'}) end,
+    sets:fold(Fun, true, Dead).
+
+pids(Tid) ->
+    Acc = couch_util:new_set(),
+    try
+        ets:foldl(fun pids_fold/2, Acc, Tid)
+    catch
+        error:badarg -> Acc
+    end.
+
+pids_fold({{Pid, _}, _, _}, Acc) when is_pid(Pid) ->
+    sets:add_element(Pid, Acc);
+pids_fold({_, _, _}, Acc) ->
+    Acc.
+
+% Private helper functions
+
+new() ->
+    Opts = [public, {write_concurrency, true}, {read_concurrency, true}],
+    Max0 = round(max_size() / erlang:system_info(wordsize) / shard_count()),
+    % Some per-table overhead for the table metadata
+    Max = Max0 + round((100 * 1024 / erlang:system_info(wordsize))),
+    #cache{tid = ets:new(?MODULE, Opts), max_size = Max}.
+
+get_cache(Term) ->
+    case persistent_term:get(?PTERM_KEY, undefined) of
+        Caches when is_tuple(Caches) ->
+            Index = erlang:phash2(Term, tuple_size(Caches)),
+            #cache{} = element(1 + Index, Caches);
+        undefined ->
+            undefined
+    end.
+
+bump_usage(Tid, Key) ->
+    % We're updating the second field incrementing it by 1 and clamping it
+    % at ?MAX_PRIORITY. We don't set the default for the update_counter
+    % specifically to avoid creating bogus entries just from updating the
+    % counter, so expect the error:badarg here sometimes.
+    UpdateOp = {2, 1, ?MAX_PRIORITY, ?MAX_PRIORITY},
+    try
+        ets:update_counter(Tid, Key, UpdateOp)
+    catch
+        error:badarg -> ok
+    end.
+
+reset_usage(Tid, Key) ->
+    % Reset the value of the usage to 0. Since max value is ?MAX_PRIORITY,
+    % subtract that and clamp it at 0. Do not provide a default since if an
+    % entry is missing we don't want to create a bogus one from this operation.
+    UpdateOp = {2, -?MAX_PRIORITY, 0, 0},
+    try
+        ets:update_counter(Tid, Key, UpdateOp)
+    catch
+        error:badarg -> ok
+    end.
+
+info(#cache{tid = Tid}) ->
+    Memory = ets:info(Tid, memory) * erlang:system_info(wordsize),
+    Size = ets:info(Tid, size),
+    {Size, Memory}.
+
+% bsr bit-shifts right so we divide by 2, then 4, 8, etc.
+%
+decay(Tid) ->
+    Head = {'$1', '$2', '$3'},
+    Cond = [{'>', '$2', 0}],
+    Result = [{{'$1', {'bsr', '$2', 1}, '$3'}}],
+    ets:select_replace(Tid, [{Head, Cond, Result}]).
```

Comment:
I think matchspec would read better here.

```erlang
MS = ets:fun2ms(
    fun({Key, Priority, Term}) when Priority > 0 ->
       % bsr bit-shifts right so we divide by 2, then 4, 8, etc.
       {Key, Priority bsr 1, Term}
    end),
ets:select_replace(Tid, MS).
```

The matchspec above produces the same query at compile time (if you would do `
-include_lib("stdlib/include/ms_transform.hrl").`).

```
8> ets:fun2ms(fun({Key, Priority, Term}) when Priority > 0 -> {Key, Priority bsr 1, Term} end).
[{{'$1','$2','$3'},
  [{'>','$2',0}],
  [{{'$1',{'bsr','$2',1},'$3'}}]}]
```

Reply:
Worked out well I think, it's much clearer to read and understand the code. Hot code loading shouldn't be an issue.